server/container/util-java8/src/test/java/org/apache/james/util/CommutativityChecker.java

```diff
@@ -0,0 +1,72 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.util;
+
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BinaryOperator;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.tuple.Pair;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.base.Preconditions;
+
+public class CommutativityChecker<T> {
+    private final Set<T> valuesToTest;
+    private final BinaryOperator<T> operationToTest;
+
+    public CommutativityChecker(Set<T> valuesToTest, BinaryOperator<T> operationToTest) {
+        Preconditions.checkNotNull(valuesToTest);
+        Preconditions.checkNotNull(operationToTest);
+        Preconditions.checkArgument(valuesToTest.size() > 1, "You must to pass more than one value to check commutativity");
+        this.valuesToTest = valuesToTest;
+        this.operationToTest = operationToTest;
+    }
+
+    public Set<Pair<T, T>> findNonCommutativeInput() {
+        return valuesToTest.stream()
+            .flatMap(this::combineWithValues)
```

Comment:
This would be messy if you mute ``valuesToTest`` you stream on it while you call method using it after word such as ``combineWithValues``.
I would make ``valuesToTest`` ImmutableSet to make that clear.
Also I would make combineWithValues static so it take two argument (value, valuesToTest). In the current state, you have to read the code of combineWithValues to understand what's going on.
Also what matter here is that you generate all possible Set of two element from the given Set. This is not directly visible without reading and understanding aaaalll the code.
I would extract this part to a method named "generateAllUnorderedPairFromSet".
Or even better we could delegate this tricky part to a library : https://github.com/dpaukov/combinatoricslib3#1-simple-combinations

Reply:
The lib looks cool <3